File: codes/perl/010.operator_nullish_coalescing.pl
use strict;
use warnings;

print("\n# JavaScript-like Nullish Coalescing Operator (??) in Perl\n");

sub pretty_json_stringify {
    my ($anything) = @_;
    use JSON;
    return JSON->new->allow_nonref->pretty->encode($anything);
}

# There's no JavaScript-like Nullish Coalescing Operator (??) in Perl.
# But, we can create our own function to mimic it in Perl.

sub nullish_coalescing {
    # JavaScript-like Nullish Coalescing Operator (??) function
    my ($value, $default_value) = @_;
    die("Required parameter not provided") unless defined($default_value);
    return !$value ? $default_value : $value;
}

my %JSON_OBJECT = (
    "foo" => {
        "bar" => "baz"
    },
    "fruits" => ["apple", "mango", "banana"]
);
print("JSON_OBJECT: ", pretty_json_stringify(\%JSON_OBJECT));

print("(JSON_OBJECT?.foo?.bar ?? \"not found\"): ", pretty_json_stringify(nullish_coalescing($JSON_OBJECT{"foo"}{"bar"}, "not found")));
# (JSON_OBJECT?.foo?.bar ?? "not found"): "baz"
print("(JSON_OBJECT?.foo?.baz ?? \"not found\"): ", pretty_json_stringify(nullish_coalescing($JSON_OBJECT{"foo"}{"baz"}, "not found")));
# (JSON_OBJECT?.foo?.baz ?? "not found"): "not found"

print("(JSON_OBJECT?.fruits?.[2] ?? \"not found\"): ", pretty_json_stringify(nullish_coalescing($JSON_OBJECT{"fruits"}[2], "not found")));
# (JSON_OBJECT?.fruits?.[2] ?? "not found"): "banana"
print("(JSON_OBJECT?.fruits?.[5] ?? \"not found\"): ", pretty_json_stringify(nullish_coalescing($JSON_OBJECT{"fruits"}[5], "not found")));
# (JSON_OBJECT?.fruits?.[5] ?? "not found"): "not found"

File: codes/ruby/010.operator_nullish_coalescing.rb
require 'json'

print("\n# JavaScript-like Nullish Coalescing Operator (??) in Ruby")

# There's no JavaScript-like Nullish Coalescing Operator (??) in Ruby.
# But, we can create our own function to mimic it in Ruby.

# JavaScript-like Nullish Coalescing Operator (??) function
nullish_coalescing = ->(anything, default_value) { ((anything == nil) ? default_value : anything) }

JSON_OBJECT = {
    "foo" => {
        "bar" => "baz"
    },
    "fruits" => ["apple", "mango", "banana"]
}
print("\nJSON_OBJECT: ", JSON.pretty_generate(JSON_OBJECT, { "indent": " " * 4 }))

print("\n# using JavaScript-like Nullish Coalescing Operator (??) function \"nullish_coalescing\"")

print("\n(JSON_OBJECT?.foo?.bar ?? \"not found\"): ", JSON.pretty_generate(nullish_coalescing.(JSON_OBJECT.dig("foo", "bar"), "not found"), { "indent": " " * 4 }))
# (JSON_OBJECT?.foo?.bar ?? "not found"): "baz"

print("\n(JSON_OBJECT?.foo?.baz ?? \"not found\"): ", JSON.pretty_generate(nullish_coalescing.(JSON_OBJECT.dig("foo", "baz"), "not found"), { "indent": " " * 4 }))
# (JSON_OBJECT?.foo?.baz ?? "not found"): "not found"

print("\n(JSON_OBJECT?.fruits?.[2] ?? \"not found\"): ", JSON.pretty_generate(nullish_coalescing.(JSON_OBJECT.dig("fruits", 2), "not found"), { "indent": " " * 4 }))
# (JSON_OBJECT?.fruits?.[2] ?? "not found"): "banana"

print("\n(JSON_OBJECT?.fruits?.[5] ?? \"not found\"): ", JSON.pretty_generate(nullish_coalescing.(JSON_OBJECT.dig("fruits", 5), "not found"), { "indent": " " * 4 }))
# (JSON_OBJECT?.fruits?.[5] ?? "not found"): "not found"
